### PR TITLE
Create an API for Handle, and rewrite calls to it

### PR DESCRIPTION
**Pull request by swankjesse**
Strictly speaking this is a more capable API than Zipline.get().
But it's more annoying to set up so I expect it to only be used
in cases when passing Flows etc. around.

### CODE REVIEW

File: zipline/src/jniTest/kotlin/app/cash/zipline/EndpointTest.kt [hunk 1]
```diff
@@ -163,4 +164,65 @@ internal class EndpointTest {
...
+    assertNull(responses.poll())
+    assertNull(requests.poll())
+  }
+
+  interface EchoServiceFactory {
+    fun create(greeting: String): Handle<EchoService>
+  }
+
+  @Test
+  fun transmitHandles() {
```
Comment (swankjesse) [inline zipline/src/jniTest/kotlin/app/cash/zipline/EndpointTest.kt:202]: This is the big payoff. We can instantiate an interface and send handles to it without names, and without directly referencing the Zipline that they connect through.

File: zipline/src/jsMain/kotlin/app/cash/zipline/ziplineJs.kt [hunk 2]
```diff
@@ -100,12 +100,12 @@ private class ZiplineJs : Zipline(), JsPlatform, CallChannel  {
...
     endpoint.set<JsPlatform>(
       name = "app.cash.zipline.jsPlatform",
-      serializersModule = ZiplineSerializersModule,
+      serializersModule = EmptySerializersModule,
       instance = this
     )
     hostPlatform = endpoint.get(
       name = "app.cash.zipline.hostPlatform",
-      serializersModule = ZiplineSerializersModule
+      serializersModule = EmptySerializersModule
```
Comment (swankjesse) [inline zipline/src/jsMain/kotlin/app/cash/zipline/ziplineJs.kt:108]: one PR I have on my queue – default parameter for `serializersModule`. It’s not implemented already because it interacts with the rewriter.

File: zipline/src/commonMain/kotlin/app/cash/zipline/internal/bridge/CallChannel.kt [hunk 3]
```diff
@@ -101,3 +97,36 @@ private object ThrowableSerializer : KSerializer<Throwable> {
...
+) : KSerializer<Handle<T>> {
+  override val descriptor = PrimitiveSerialDescriptor("Handle", PrimitiveKind.STRING)
+
+  override fun serialize(encoder: Encoder, value: Handle<T>) {
+    val name = endpoint.generateName()
+    if (value is InboundHandle<T>) {
+      value.connect(endpoint, name)
+      encoder.encodeString(name)
+    } else {
+      error("serializing an outbound handle is not implemented")
```
Comment (swankjesse) [inline zipline/src/commonMain/kotlin/app/cash/zipline/internal/bridge/CallChannel.kt:122]: There’s a fancy implementation possible here. . . if you send a Flow back to its origin we can send it back by reference and not by value.

File: zipline/src/commonMain/kotlin/app/cash/zipline/Handle.kt [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+      include(outboundEndpoint.builtInSerializersModule)
+      include(outboundBridge.serializersModule)
+    }
+
+    val context = OutboundBridge.Context(name, serializersModule, outboundEndpoint)
+    return outboundBridge.create(context)
+  }
+
+  override fun close() {
+    // TODO(jwilson): support detaching outbound handles.
```
Comment (swankjesse) [inline zipline/src/commonMain/kotlin/app/cash/zipline/Handle.kt:103]: the only reason I haven’t done this already is it needs a bridged call to perform the removal.

File: zipline/src/commonMain/kotlin/app/cash/zipline/Handle.kt [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    }
+
+    val context = InboundBridge.Context(serializersModule)
+    val result = inboundBridge.create(context)
+    endpoint.inboundHandlers[name] = result
+    return result
+  }
+
+  override fun get(outboundBridge: OutboundBridge<T>): T {
+    // TODO(jwilson): change generated subtypes to populate `T` on the super call.
```
Comment (swankjesse) [inline zipline/src/commonMain/kotlin/app/cash/zipline/Handle.kt:63]: This is an important ergonomics thing. . . if you’re writing a test you’ll have an `InboundHandle` and will want to call `get()` on it.

File: zipline/src/commonMain/kotlin/app/cash/zipline/Handle.kt [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+ * limitations under the License.
+ */
+package app.cash.zipline
+
+import app.cash.zipline.internal.bridge.Endpoint
+import app.cash.zipline.internal.bridge.InboundBridge
+import app.cash.zipline.internal.bridge.InboundCallHandler
+import app.cash.zipline.internal.bridge.OutboundBridge
+import kotlinx.serialization.modules.SerializersModule
+import okio.Closeable
```
Comment (JakeWharton) [inline zipline/src/commonMain/kotlin/app/cash/zipline/Handle.kt:23]: Annoying

File: zipline/src/commonMain/kotlin/app/cash/zipline/Handle.kt [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+    throw IllegalStateException("get() on inbound handlers not yet implemented")
+  }
+
+  override fun close() {
+    val endpoint = this.endpoint
+    val name = this.name
+    if (endpoint != null && name != null) {
+      val removed = endpoint.inboundHandlers.remove(name)
+      this.endpoint = null
+      this.name = null
```
Comment (JakeWharton) [inline zipline/src/commonMain/kotlin/app/cash/zipline/Handle.kt:73]: Does this mean you can somehow connect more than once?

**Review by JakeWharton — APPROVED** [review]

File: zipline/src/commonMain/kotlin/app/cash/zipline/Handle.kt (hunk 7 above)
Comment (swankjesse) [reply zipline/src/commonMain/kotlin/app/cash/zipline/Handle.kt:73]: I was thinking about that. I don’t think that makes sense and I should explicitly forbid it. Lemme implement that in a follow round with lots of new tests and edge cases.
